src/deadline_worker_agent/linux/capabilities.py

```diff
@@ -0,0 +1,215 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+
+"""This module contains code for interacting with Linux capabilities.
+
+See https://man7.org/linux/man-pages/man7/capabilities.7.html for details on this Linux kernel
+feature.
+"""
+
+import ctypes
+import os
+import sys
+from logging import getLogger
+from ctypes.util import find_library
+from functools import cache
+from typing import Any, Tuple, TYPE_CHECKING
+
+
+logger = getLogger(__name__)
+
+
+# Capability sets
+CAP_EFFECTIVE = 0
+CAP_PERMITTED = 1
+CAP_INHERITABLE = 2
+
+# Capability bit numbers
+CAP_KILL = 5
+
+# Values for cap_flag_value_t arguments
+CAP_CLEAR = 0
+CAP_SET = 1
+
+cap_flag_t = ctypes.c_int
+cap_flag_value_t = ctypes.c_int
+cap_value_t = ctypes.c_int
+
+
+class UserCapHeader(ctypes.Structure):
+    _fields_ = [
+        ("version", ctypes.c_uint32),
+        ("pid", ctypes.c_int),
+    ]
+
+
+class UserCapData(ctypes.Structure):
+    _fields_ = [
+        ("effective", ctypes.c_uint32),
+        ("permitted", ctypes.c_uint32),
+        ("inheritable", ctypes.c_uint32),
+    ]
+
+
+class Cap(ctypes.Structure):
+    _fields_ = [
+        ("head", UserCapHeader),
+        ("data", UserCapData),
+    ]
+
+
+if TYPE_CHECKING:
+    cap_t = ctypes._Pointer[Cap]
+    cap_flag_value_ptr = ctypes._Pointer[cap_flag_value_t]
+    cap_value_ptr = ctypes._Pointer[cap_value_t]
+    ssize_ptr_t = ctypes._Pointer[ctypes.c_ssize_t]
+else:
+    cap_t = ctypes.POINTER(Cap)
+    cap_flag_value_ptr = ctypes.POINTER(cap_flag_value_t)
+    cap_value_ptr = ctypes.POINTER(cap_value_t)
+    ssize_ptr_t = ctypes.POINTER(ctypes.c_ssize_t)
+
+
+def _cap_set_proc_err_check(
+    result: ctypes.c_int,
+    func: Any,
+    args: Tuple[Any, ...],
+) -> ctypes.c_int:
+    if result != 0:
+        errno = ctypes.get_errno()
+        raise OSError(errno, os.strerror(errno))
+    return result
+
+
+def _cap_get_proc_err_check(
+    result: cap_t,
+    func: Any,
+    args: Tuple[cap_t, cap_flag_t, ctypes.c_int, cap_value_ptr, cap_flag_value_t],
+) -> cap_t:
+    if not result:
+        errno = ctypes.get_errno()
+        raise OSError(errno, os.strerror(errno))
+    return result
+
+
+def _cap_to_text_errcheck(
+    result: ctypes.c_char_p,
+    func: Any,
+    args: Tuple[cap_t, ssize_ptr_t],
+) -> ctypes.c_char_p:
+    if not result:
+        errno = ctypes.get_errno()
+        raise OSError(errno, os.strerror(errno))
+    return result
+
+
+def _cap_get_flag_errcheck(
+    result: ctypes.c_int, func: Any, args: Tuple[cap_t, cap_value_t, cap_flag_t, cap_flag_value_ptr]
+) -> ctypes.c_int:
+    if result != 0:
+        errno = ctypes.get_errno()
+        raise OSError(errno, os.strerror(errno))
+    return result
+
+
+@cache
+def _get_libcap() -> ctypes.CDLL:
+    if not sys.platform.startswith("linux"):
+        raise OSError(f"libcap is only available on Linux, but found platform: {sys.platform}")
+
+    libcap = ctypes.CDLL(find_library("cap"), use_errno=True)
```

Comment:
Same issue as the `openjd-sessions` PR. We're assuming that the lib is always present. We modified the other change to account for it not being present.

Reply:
I've added error handling for when `libcap` cannot be found. On Amazon Linux, it appears that systemd depends on libcap:

```sh
$ dnf repoquery -q --installed --whatrequires libcap
chrony-0:4.3-1.amzn2023.0.4.x86_64
coreutils-0:8.32-30.amzn2023.0.3.x86_64
iproute-0:6.10.0-319.amzn2023.0.1.x86_64
iputils-0:20210202-2.amzn2023.0.4.x86_64
kernel-tools-0:6.1.115-126.197.amzn2023.x86_64
libcap-devel-0:2.48-2.amzn2023.0.3.x86_64
nfs-utils-1:2.5.4-2.rc3.amzn2023.0.3.x86_64
rpm-build-libs-0:4.16.1.3-29.amzn2023.0.6.x86_64
rpm-libs-0:4.16.1.3-29.amzn2023.0.6.x86_64
rpm-plugin-selinux-0:4.16.1.3-29.amzn2023.0.6.x86_64
rpm-plugin-systemd-inhibit-0:4.16.1.3-29.amzn2023.0.6.x86_64
rpm-sign-libs-0:4.16.1.3-29.amzn2023.0.6.x86_64
systemd-0:252.23-2.amzn2023.x86_64
systemd-libs-0:252.23-2.amzn2023.x86_64
systemd-pam-0:252.23-2.amzn2023.x86_64
```

systemd is used to launch the worker agent on boot and restart on a crash. We also configure the worker agent to be given `CAP_KILL` in the systemd unit generated by `install-deadline-worker`.

On this basis, we decided to log a warning when the library is not found and move on. Those running the worker agent on Linux with `CAP_KILL` but without `libcap` available is not officially supported.